### PR TITLE
feat: add tree entities update and delete logic

### LINKED ISSUE #193: Remove entity from TreeRepository (Closure table)
Does the `remove() `method work for Closure tables? When I tried to remove a leaf node from a Category model, I got the following error message (using typeorm 0.0.6, typescript, mysql 5):

`ER_ROW_IS_REFERENCED_2: Cannot delete or update a parent row: a foreign key constraint fails (`exampledb`.`category_closure`, CONSTRAINT `fk_11da9decc6d053c89b14825abe2` FOREIGN KEY (`descendant`) REFERENCES `category` (`id`))`

I checked the samples, but they only cover persisting and retrieving items from a TreeRepository.

### LINKED ISSUE #2032: Tree closure table is not updated on parent update
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[x] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Closure table is not updated when previously created entity updates its parent.

```import 'reflect-metadata';
import * as path from 'path';
import { createConnection, ConnectionOptions, getTreeRepository } from 'typeorm';
import { PrimaryGeneratedColumn, Column, Tree, TreeChildren, TreeParent, Entity } from 'typeorm';

@Entity()
@Tree("closure-table")
export class TestEntity {
	@PrimaryGeneratedColumn() id: number;
	@TreeChildren()
	children: TestEntity[];
	@TreeParent() parent: TestEntity;
}

var connectionOptions: ConnectionOptions = {
    type: 'sqlite',
    database: path.join(__dirname, 'test.db'),
    entities: [ TestEntity ],
    synchronize: true,
    logging: true,
    dropSchema: true
};

createConnection(connectionOptions).then(async (c) => {
    var repo = getTreeRepository(TestEntity);
    
    var parent1 = new TestEntity;
    await repo.save(parent1);
    var parent2 = new TestEntity;
    await repo.save(parent2);
    var child = new TestEntity;

    child.parent = parent1;
    await repo.save(child);
    console.log(...await repo.query(`SELECT * FROM ${repo.metadata.closureJunctionTable.name}`));
    
    child.parent = parent2;
    await repo.update(child.id, child);
    console.log(...await repo.query(`SELECT * FROM ${repo.metadata.closureJunctionTable.name}`));
});
```

### LINKED ISSUE #2418: incorrect update Tree
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

tree-entities ("materialized-path") - incorrect update query
```
[Node] query: START TRANSACTION
[Node] query: UPDATE `auth`.`role` SET `parentId` = ? WHERE `id` = ? -- PARAMETERS: [1,3]
[Node] query: COMMIT
```
- query not updatet field `mpath`

1 - Create Entity
```ts
import {
    Entity,
    Tree,
    Column,
    PrimaryGeneratedColumn,
    TreeChildren,
    TreeParent,
    TreeLevelColumn,
    ManyToMany,
    JoinTable
} from "typeorm";

@Entity({ database: "auth" })
@Tree("materialized-path")
export class Role {
    @PrimaryGeneratedColumn() id: number;

    @Column()
    name: string;

    @Column({
        nullable: true
    })
    description: string;

    @TreeChildren() children: Role[];

    @TreeParent() parent: Role;
}

```

2 - Create 3 Record
```ts
        const treeRepositoryRole = getTreeRepository(Role, "auth");
        const role1 = new Role(); // id 1
        role1.name = 'admin'
        const role2 = new Role(); // id 2
        role2.name = 'Subadmin'
        const role3 = new Role(); // id 3
        role3.name = 'SubSubadmin'
        await repositoryRole.save(role1);
        await repositoryRole.save(role2);
        await repositoryRole.save(role3);
```

3 - update 1 record
```ts
   const treeRepositoryRole = getTreeRepository(Role, "auth");
   const role = await repositoryRole.findOne(3);
   const roleParent = await repositoryRole.findOne(1);
   role.parent = roleParent;
   await treeRepositoryRole.save(role);
``` 
4 - console logs
```
[Node] query: START TRANSACTION
[Node] query: UPDATE `auth`.`role` SET `parentId` = ? WHERE `id` = ? -- PARAMETERS: [1,3]
[Node] query: COMMIT
```
query not updated `mpath` field

### LINKED ISSUE #3691: I am having trouble using closure-table
**Issue type:**

- [x] bug report

**Database system/driver:**

- [x] `mysql` / `mariadb`

**TypeORM version:**

- [x] `latest`

**Steps to reproduce or a small repository showing the problem:**

```typescript
// Entity
import {Column, PrimaryGeneratedColumn,Tree,TreeChildren,TreeParent} from "typeorm";
@Tree("closure-table")
export class Category {
    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    name: string;
    
    @TreeParent()
    parentCategory: Category;

    @TreeChildren({ cascade: true })
    childCategories: Category[];
}

// code
public async delete(data: number): Promise<Category> {
        const result = await this.categoryRepository.findOne({ id: data });
        this.logger.warn(result);
        return await this.categoryRepository.remove(result);
}

// message
`ER_ROW_IS_REFERENCED_2: Cannot delete or update a parent row: a foreign key constraint fails (`lee`.`category_closure`, CONSTRAINT `FK_4aa1348fc4b7da9bef0fae8ff48` FOREIGN KEY (`id_ancestor`) REFERENCES `category` (`id`))`
```
**Steps to reproduce or a small repository showing the problem:**

### LINKED ISSUE #3988: How to delete in Closure table of Tree Entities?
**Issue type:**
[x] question

**Database system/driver:**
[x] `mysql` / `mariadb`

**TypeORM version:**
[x] `@next`

```
@Column({
    length: 500,
    default: '',
  })
  name: string

@TreeParent()
  parent: Comment

  @TreeChildren({
    cascade: true,
  })
  children: Comment[]
```
above code, I want to delete a record, but error report: ` a foreign key constraint fails...`，and how to delete in this case ?

### LINKED ISSUE #4174: move entities in nested-set
**Issue type:**

[x] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
Hi, we are using typeorm in some of our projects and we are mostly happy with it ;-). However, we found, that in nested-set tree entities there is no moving functionality. One could load an entity, replace the parent with another entity and the update query would correctly update the parentId but would not touch the nsleft and nsright column. As my (admittedly limited) understanding is, these columns should at least be updated when the moved entity has children.

My question is, does this break any tree functionality like the findDescendantsTree method or can it safely be ignored?
Should there be some update to nsleft and nsright similar to the [query that is fired before inserts](https://github.com/typeorm/typeorm/blob/0.2.17/src/persistence/tree/NestedSetSubjectExecutor.ts#L24) in nested sets?

I looked at the code and the tests for tree repositories and entities and saw, that there is an "implement moving todo". As I understand, new features are currently not implemented. Can anybody shed some light on this?

Edit: This is possibly related to https://github.com/typeorm/typeorm/pull/3556 and https://github.com/typeorm/typeorm/issues/2418.

### LINKED ISSUE #4353: How to update Tree with closure table?
**Issue type:**

[ X] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ X] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

<!--
    To answer those questions you need to put "x" inside the square brackets, for example:
    [x] `mysql`
    [ ] `postgres`

    Also, please format your code properly (by taking code blocks into ```ts .... ```)
!>

Hi,

Has anyone figured out how to update a tree stored in closure table?

I followed the example given in the documentation but there seems to be no obvious way to update a node or for example move a child node to another parent.

Is this implemented or not?

I have tried to delete the category from the category table and the categories_closure table in order to re-insert it with a different parent.  try to delete it like this

```
    await getManager()
      .createQueryBuilder()
      .delete()
      .from('categories_closure')             
      .where('"id_ancestor" = :id', { id })
      .execute();
    
    await this.repo.delete(id);

```

which reasonably produces error:

```
error: update or delete on table "categories" violates foreign key constraint "FK_51fff5114cc41723e8ca36cf227" on table "categories_closure"
```

Thank you.

### LINKED ISSUE #5268: Is there a way to change default CONSTRAINT for tree closure table? I use postgres.
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

1. create a table with `@Tree('closure-table')` decorator
2. insert at least one row to created table
3. add at least one child to tree
4. try to remove parrent row

Expected:
Parent row removed with all his children

Actual:
Parent row could not be removed, because of CONSTRAINT restcition of closure-table

### LINKED ISSUE #6118: Tree  closure-table doesn't update for my code
**Issue type:**

[x] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

### Tree closure-table not update

The normal  data :
I have three items for organization table
```
id          title          parentId
1           1               null
2           2              null
3           1-1            1
```
the organization_closure like this
```
id_ancestor    id_descendant
1                      1
1                      3
2                      2
3                      3
```
the sql run normal
first item's sql:
```
query: INSERT INTO `organization`(`id`, `title`, `mark`, `managerId`, `parentId`) VALUES (DEFAULT, ?, ?, DEFAULT, DEFAULT) -- PARAMETERS: ["1","1"]
query: INSERT INTO `organization_closure`(`id_ancestor`, `id_descendant`) VALUES (?, ?) -- PARAMETERS: [1,1]
```

second item's sql
```
query: INSERT INTO `organization`(`id`, `title`, `mark`, `managerId`, `parentId`) VALUES (DEFAULT, ?, ?, DEFAULT, DEFAULT) -- PARAMETERS: ["2","2"]
query: INSERT INTO `organization_closure`(`id_ancestor`, `id_descendant`) VALUES (?, ?) -- PARAMETERS: [2,2]
```

third item’s sql
```
query: INSERT INTO `organization`(`id`, `title`, `mark`, `managerId`, `parentId`) VALUES (DEFAULT, ?, ?, DEFAULT, ?) -- PARAMETERS: ["1-1","1-1",1]
query: INSERT INTO `organization_closure`(`id_ancestor`, `id_descendant`) VALUES (?, ?) -- PARAMETERS: [3,3]
query: INSERT INTO `organization_closure` (`id_ancestor`, `id_descendant`) SELECT `id_ancestor`, ? FROM `organization_closure` WHERE `id_descendant` = ? -- PARAMETERS: [3,1]
```

The question is , I want to move the third item as second item's child
```
    .....
    const orgQuery = queryRunner.manager.getTreeRepository(Organization)

    const cOrg = await orgQuery.findOne({ id: params.id })
    const newOrgParent = await orgQuery.findOne({ id: params.parent })

    Object.assign(cOrg, params)
    cOrg.parent = newOrgParent
    await orgQuery.update(cOrg.id, cOrg)
    ....
```
Run the code i can't see the organization_closure table's sql? why? just has organization's sql~
```
query: UPDATE `organization` SET `id` = ?, `title` = ?, `mark` = ?, `managerId` = ?, `parentId` = ? WHERE `id` IN (?) -- PARAMETERS: [3,"1-1","1-1",null,2,3]
```
and now table data like this
```
id          title          parentId
1           1               null
2           2               null
3           1-1             2
```
the organization_closure like this
```
id_ancestor    id_descendant
1                      1
1                      3
2                      2
3                      3
```

where my question is?

### PR DESCRIPTION
**Pull request by mateussilva92**
Allows tree entities relations to be updated and deleted via the save function from the Repository

Closes: #7155

### Description of change
Main changes are done to main `SubjectExecutor.ts` and all the tree executors.

I added logic to all the tree executors to allow tree relation changes via the update function and also added the delete logic to `NestedSetSubjectExecutor.ts`.
On the `SubjectExecutor.ts` the biggest change done was to run the **nested set** updates in series instead of parallel just like all the other entities updates are done. The reason for that is explained in #7155.
Enable onDelete option for TreeParent in all RDBMS except for mssql.
Also cascade delete for closure junction table is also enabled to all RDBMS and as custom code to had the same behaviour in mssql.

All the other file changes are either small format changes, code cleanup or some changes that allow the main new feature to work.

Fixes #193
Fixes #2032
Fixes #2418
Fixes #3691
Fixes #3988
Fixes #4174
Fixes #4353
Fixes #5268
Fixes #6118

Duplicated by #3556

_Note:_
The development also prevents the creation of more than one root entity for **nested sets** I'm not sure if that's an approach that you would like to have or not.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/metadata-builder/ClosureJunctionEntityMetadataBuilder.ts [hunk 1]
```diff
@@ -110,20 +110,21 @@ export class ClosureJunctionEntityMetadataBuilder {
         }
 
         // create junction table foreign keys
+        // Note: There is only a single CASCADE because of mssql not supporting multi cascade paths
+        //       However this should not cause any problems since the both foreign keys point to the same column on the same table
```
Comment (nebkat) [inline src/metadata-builder/ClosureJunctionEntityMetadataBuilder.ts:114]: There is access to `this.connection.driver` here, should there be a check for `instanceof SqlServerDriver`?

Comment (mateussilva92) [reply src/metadata-builder/ClosureJunctionEntityMetadataBuilder.ts:114]: We do have access to the driver yes.. I would need to add some extra logic to the delete inside the subject executor.

I also found that the changes that I did on `TreeParent.ts` to allow deleting the parents via cascade are also causing the same problem on sql server so I'll probably need to revert that change and add custom delete logic for every time of tree entity.

**Comment by boydaihungst** [comment]
I got ```Cannot read property 'find' of undefined``` error when used with  TableInheritance.

```
@ObjectType({ isAbstract: true })
@Entity('file_folder')
@TableInheritance({ column: { type: 'varchar', name: 'type' } })
@Tree('closure-table')
export class FileFolderBase extends CustomBaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ type: 'nvarchar', length: 255, nullable: false })
  @MaxLength(255, { message: DbError.FIELD_LENGTH_MAX })
  name: string;

  @TreeChildren({ cascade: ['soft-remove', 'remove', 'recover'] })
  childrens: FileFolderBase[];

  @TreeParent({ onDelete: 'CASCADE' })
  parentFolder: FileFolderBase;
}
```

```
@ChildEntity()
@Tree('closure-table')
export class Folder extends FileFolderBase {
  @Column({ type: 'boolean', nullable: false, default: false })
  isRoot?: boolean;

  @Column({ type: 'varchar', length: 45, nullable: true })
  color?: string;
}
```

This is how i create and save new folder.
```
    const result = await this.folderRepository.save(
      this.folderRepository.create({
        name: `ROOT/${userId}`,
        isRoot: true,
      }),
    );
```
Error at line 70 
https://github.com/typeorm/typeorm/blob/1358dfbb1a88ad3b3fc1465e3d5296d75ddd0732/src/persistence/tree/ClosureSubjectExecutor.ts#L67-L73
And if I check undefined its works normally
```
        if (!entity) {
            // if entity was attached via children
            const chilrens = subject.metadata.treeChildrenRelation!.getEntityValue(
                parent
            );
            if (!!chilrens)
                entity = chilrens!.getEntityValue(parent).find((child: any) => {
                    return Object.entries(subject.identifier!).every(
                        ([key, value]) => child[key] === value
                    );
                });
        }
```

**Comment by mateussilva92** [comment]
@boydaihungst
Going to fix that later on and add some new tests to prevent that from happening again. Thanks for the feedback.

**Comment by mateussilva92** [comment]
@boydaihungst
I tried to replicate the issue on my side but I was not able to, can you please tell me what database are you using and also share the code that you use to update your entity since the one you share is to do the inicial save which should not go through the update flow.

**Comment by boydaihungst** [comment]
@mateussilva92 
Sorry for the lack of information. I use Mysql. And I updated the comment below. I think the problem is I use ```cascade: true``` for property ``` statistic ```
 ```
statistic: { 
    downloaded: BigInt(0),
    viewed: BigInt(0),
}
```

**Comment by mateussilva92** [comment]
@boydaihungst
Still can't replicate even after the update you gave.
I don't think it's the `cascade: true` on the `statistic` because that entity is not even a `TreeEntity`.
Is there any change you could provide a minimal repo with all the code to help me replicate the issue?

**Comment by boydaihungst** [comment]
@mateussilva92 
Here is minial repo
[https://github.com/boydaihungst/typeorm-tree-error](https://github.com/boydaihungst/typeorm-tree-error)

**Comment by mateussilva92** [comment]
@boydaihungst
Thanks for the repo it really helped. The issue is now fixed.

**Comment by kauandotnet** [comment]
@pleerock 
Please, check this! I'm working for a project that need this.

Thank you!

**Comment by phil294** [comment]
Hi @mateussilva92, thanks for doing this! I've been using it for a few weeks in production with success. There is one issue however: Setting a parent property to `null` and saving the entity afterwards leads to ```error: "CannotAttachTreeChildrenEntityError: Cannot attach entity \"Post\" to its parent. Please make sure parent is saved in the database before saving children nodes.
    at new CannotAttachTreeChildrenEntityError (/project/src/error/CannotAttachTreeChildrenEntityError.ts:8:9)
    at /project/src/persistence/tree/ClosureSubjectExecutor.ts:251:23
    at Array.map (<anonymous>)
    at ClosureSubjectExecutor.<anonymous> (/project/src/persistence/tree/ClosureSubjectExecutor.ts:246:88)
    at step (/project/node_modules/tslib/tslib.js:141:27)
    at Object.next (/project/node_modules/tslib/tslib.js:122:57)
    at /project/node_modules/tslib/tslib.js:115:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/project/node_modules/tslib/tslib.js:111:16)
    at ClosureSubjectExecutor.insertClosureEntry (/project/node_modules/typeorm/persistence/tree/ClosureSubjectExecutor.js:242:24)"```
Setting the `parentId` field to `null` instead does not throw an error, but competely fails to drop the relevant closure table entries. So there seem to be two bugs here.
Does this make sense / am I overseeing something?
If it *should* work, please tell me so I can create a reproduction repository.

**Comment by mateussilva92** [comment]
Hey @phil294 thanks for the good catch and for the awesome feedback with the repo for the bug.
I'll try to fix the issue on the next few days.

File: src/driver/sqlserver/SqlServerQueryRunner.ts [hunk 2]
```diff
@@ -1288,6 +1289,15 @@ export class SqlServerQueryRunner extends BaseQueryRunner implements QueryRunner
      */
     async createForeignKey(tableOrName: Table|string, foreignKey: TableForeignKey): Promise<void> {
         const table = tableOrName instanceof Table ? tableOrName : await this.getCachedTable(tableOrName);
+        
+        let metadata: EntityMetadata | undefined = undefined;
+        try {
+            metadata = this.connection.getMetadata(table.name);
+        } catch (err) { } // Ignore error and continue
```
Comment (AlexMesser) [inline src/driver/sqlserver/SqlServerQueryRunner.ts:1296]: `try ...catch` can be replaced with `hasMetadata(table.name)` condition

Comment (mateussilva92) [reply src/driver/sqlserver/SqlServerQueryRunner.ts]: Thanks for the catch, fixed.

**Comment by AlexMesser** [comment]
thank you for contribution!

**Pull request merged**